File: DSA/Dart/Sorting/mergeSort.dart
List<int> mergeSort(List<int> list) {
  // Base case: if the list is empty or have a one element,
  if (list.length <= 1) {
    return list;
  }
  // Spilt it into halves
  int middle = list.length ~/ 2;
  List<int> left = list.sublist(0, middle);
  List<int> right = list.sublist(middle);
  // Recursive sort halves
  left = mergeSort(left);
  right = mergeSort(right);

  // Merge the sort halves
  return merge(left, right);
}

List<int> merge(List<int> left, List<int> right) {
  List<int> result = [];
  int i = 0;
  int j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      result.add(left[i]);
      i++;
    } else {
      result.add(right[j]);
      j++;
    }
  }
  // Remaning the left element if any
  while (i < left.length) {
    result.add(left[i]);
    i++;
  }
  while (j < right.length) {
    result.add(right[j]);
    j++;
  }
  return result;
}

void main() {
  List<int> unsortedList = [6, 4, 2, 1, 9, 8, 3, 5];
  List<int> sortedList = mergeSort(unsortedList);
  print(sortedList);
}

File: DSA/Python/Sorting/mergeSort.py
def margeSort(list):
    length = len(list)
    # Base case
    if length <=  1:
        return list
    middle = length // 2
    
    left_half = list[:middle]
    right_half = list[middle:]
    
    left_half = margeSort(left_half)
    right_half = margeSort(right_half)
    
    return merge(left_half, right_half)

def merge(left, right): 
    result = []
    i = 0
    j = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    while i < len(left):
        result.append(left[i])
        i += 1
    
    while j < len(right):
        result.append(right[j])
        j += 1
        
    return result     

size = int(input("Enter the size of the list: "))

element = [int(input("Enter the element: ")) for i in range(size)]

sortedList = margeSort(element)
print(sortedList)

